src/server.rs

```diff
@@ -54,22 +62,56 @@ impl<S: Service, C: ServiceEndpoint<S>> RpcServer<S, C> {
 /// Sink and stream are independent, so you can take the channel apart and use
 /// them independently.
 #[derive(Debug)]
-pub struct RpcChannel<S: Service, C: ServiceEndpoint<S>> {
+pub struct RpcChannel<S: Service, C: ServiceEndpoint<S>, S2: Service = S> {
     /// Sink to send responses to the client.
     pub send: C::SendSink,
     /// Stream to receive requests from the client.
     pub recv: C::RecvStream,
-    /// Phantom data to make the type parameter `S` non-instantiable.
-    p: PhantomData<S>,
+    /// Mapper to map between S and S2
+    map: Arc<dyn MapService<S, S2>>,
```

Comment:
Ah, forget about this, this does not work because S != S2, right?

Reply:
Yup. I tried to make it optional for the case where S = S2, but failed so far. I tried with an enum `MaybeDynMapper` but couldn't get the bounds right, not sure if there is a way. I think it would need conditional impls based on bounds, which is not possible I think.